### PR TITLE
feat: Warn on outdated plugins.

### PR DESCRIPTION
**Pull request by marianogappa**
On sync, it looks into the source and destination plugin versions and compares them against the latest version according to the public `GetPluginWithResponse` API call. Then, if current is lesser than latest, it warns:

```bash
$ go run . test-connection ...all.yml --log-console
2024-10-16T11:33:38Z INF Loading spec(s) args=["...all.yml"] invocation_id=3236492c-0bf8-4823-a0f1-6a0d21dc2b0c module=cli
2024-10-16T11:33:40Z WRN Plugin is outdated, consider upgrading to the latest version. invocation_id=3236492c-0bf8-4823-a0f1-6a0d21dc2b0c latest_version=27.23.1 module=cli plugin=aws url=https://hub.cloudquery.io/plugins/source/cloudquery/aws using_version=27.23.0
2024-10-16T11:33:41Z WRN Plugin is outdated, consider upgrading to the latest version. invocation_id=3236492c-0bf8-4823-a0f1-6a0d21dc2b0c latest_version=8.6.2 module=cli plugin=postgresql url=https://hub.cloudquery.io/plugins/destination/cloudquery/postgresql using_version=8.0.7
```

```bash
$ go run . install ...all.yml --log-console
Command "install" is deprecated, use `cloudquery plugin install` instead
2024-10-16T11:34:30Z INF Loading spec(s) args=["...all.yml"] invocation_id=a14ae9aa-fe4c-40ce-8dc1-3dfae9a38061 module=cli
2024-10-16T11:34:32Z WRN Plugin is outdated, consider upgrading to the latest version. invocation_id=a14ae9aa-fe4c-40ce-8dc1-3dfae9a38061 latest_version=27.23.1 module=cli plugin=aws url=https://hub.cloudquery.io/plugins/source/cloudquery/aws using_version=27.23.0
2024-10-16T11:34:32Z WRN Plugin is outdated, consider upgrading to the latest version. invocation_id=a14ae9aa-fe4c-40ce-8dc1-3dfae9a38061 latest_version=8.6.2 module=cli plugin=postgresql url=https://hub.cloudquery.io/plugins/destination/cloudquery/postgresql using_version=8.0.7
```

```bash
$ go run . sync ...all.yml --log-console
2024-10-16T11:34:46Z INF Loading spec(s) args=["...all.yml"] invocation_id=a5975cc7-5625-484c-97b0-b14f5d6f5658 module=cli
2024-10-16T11:34:48Z WRN Plugin is outdated, consider upgrading to the latest version. invocation_id=a5975cc7-5625-484c-97b0-b14f5d6f5658 latest_version=27.23.1 module=cli plugin=aws url=https://hub.cloudquery.io/plugins/source/cloudquery/aws using_version=27.23.0
2024-10-16T11:34:48Z WRN Plugin is outdated, consider upgrading to the latest version. invocation_id=a5975cc7-5625-484c-97b0-b14f5d6f5658 latest_version=8.6.2 module=cli plugin=postgresql url=https://hub.cloudquery.io/plugins/destination/cloudquery/postgresql using_version=8.0.7
```

```bash
$ go run . tables ...all.yml --log-console
2024-10-16T11:35:28Z INF Loading spec(s) args=["...all.yml"] invocation_id=e9029046-92fd-45ca-b8e9-d09e7e3918f0 module=cli
2024-10-16T11:35:30Z WRN Plugin is outdated, consider upgrading to the latest version. invocation_id=e9029046-92fd-45ca-b8e9-d09e7e3918f0 latest_version=27.23.1 module=cli plugin=aws url=https://hub.cloudquery.io/plugins/source/cloudquery/aws using_version=27.23.0
```

```bash
$ go run . migrate ...all.yml --log-console
2024-10-16T11:36:00Z INF Loading spec(s) args=["...all.yml"] invocation_id=ce95f355-8fd8-45ba-8ab3-ac417d34eb54 module=cli
2024-10-16T11:36:01Z WRN Plugin is outdated, consider upgrading to the latest version. invocation_id=ce95f355-8fd8-45ba-8ab3-ac417d34eb54 latest_version=27.23.1 module=cli plugin=aws url=https://hub.cloudquery.io/plugins/source/cloudquery/aws using_version=27.23.0
2024-10-16T11:36:01Z WRN Plugin is outdated, consider upgrading to the latest version. invocation_id=ce95f355-8fd8-45ba-8ab3-ac417d34eb54 latest_version=8.6.2 module=cli plugin=postgresql url=https://hub.cloudquery.io/plugins/destination/cloudquery/postgresql using_version=8.0.7
```

Init uses latest, so it doesn't need this.

### CODE REVIEW

File: cli/cmd/sync.go [hunk 1]
```diff
@@ -233,6 +234,11 @@ func sync(cmd *cobra.Command, args []string) error {
 		if err != nil {
 			return enrichClientError(managedplugin.Clients{}, []bool{source.RegistryInferred()}, err)
 		}
+		latestVersion, err := sourcePluginClient.FindLatestPluginVersion(ctx, managedplugin.PluginSource)
+		if err == nil && source.Version != "" && latestVersion != "" && semver.IsLess(source.Version, latestVersion) {
+			log.Warn().Str("source", source.Name).Msgf("Your source plugin is out of date, you are using version %s, and the latest version is %s. Consider upgrading.", source.Version, latestVersion)
```
Comment (disq) [inline cli/cmd/sync.go:239]: Could (or should) we also include a link to the hub? Like:

```suggestion
			log.Warn().Str("source", source.Name).Str("url", "https://hub.cloudquery.io/plugins/source/" + source.Path).Msgf("Your source plugin is out of date, you are using version %s, and the latest version is %s. Consider upgrading.", source.Version, latestVersion)
```

File: cli/cmd/sync.go [hunk 2]
```diff
@@ -276,6 +282,10 @@ func sync(cmd *cobra.Command, args []string) error {
 		if err != nil {
 			return enrichClientError(managedplugin.Clients{}, []bool{destination.RegistryInferred()}, err)
 		}
+		latestVersion, err := destPluginClient.FindLatestPluginVersion(ctx, managedplugin.PluginDestination)
+		if err == nil && destination.Version != "" && latestVersion != "" && semver.IsLess(destination.Version, latestVersion) {
+			log.Warn().Str("source", destination.Name).Msgf("Your destination plugin is out of date, you are using version %s, and the latest version is %s. Consider upgrading.", destination.Version, latestVersion)
```
Comment (disq) [inline cli/cmd/sync.go:287]:
```suggestion
			log.Warn().Str("destination", destination.Name).Msgf("Your destination plugin is out of date, you are using version %s, and the latest version is %s. Consider upgrading.", destination.Version, latestVersion)
```

Comment (disq) [inline cli/cmd/sync.go:287]: Including a link to the hub would look like:
```suggestion
			log.Warn().Str("source", destination.Name).Str("url", "https://hub.cloudquery.io/plugins/destination/" + destination.Path).Msgf("Your destination plugin is out of date, you are using version %s, and the latest version is %s. Consider upgrading.", destination.Version, latestVersion)
```

**Review by erezrokah — COMMENTED** [review]
We should add this to transformer plugins too

File: cli/cmd/sync.go (hunk 2 above)
Comment (marianogappa) [reply cli/cmd/sync.go]: Fixed

File: cli/cmd/sync.go (hunk 1 above)
Comment (marianogappa) [reply cli/cmd/sync.go]: Fixed.

**Comment by marianogappa** [comment]
Added to transformer plugins too.

File: cli/cmd/sync.go [hunk 3]
```diff
@@ -233,6 +234,11 @@ func sync(cmd *cobra.Command, args []string) error {
 		if err != nil {
 			return enrichClientError(managedplugin.Clients{}, []bool{source.RegistryInferred()}, err)
 		}
+		latestVersion, err := sourcePluginClient.FindLatestPluginVersion(ctx, managedplugin.PluginSource)
```
Comment (erezrokah) [inline cli/cmd/sync.go:237]: I think `managedplugin.NewClient(ctx, managedplugin.PluginSource, cfg, opts...)` downloads the plugins binary first which is not needed and will fail it auth is not provided I believe

File: cli/internal/semver/semver.go [hunk 4]
```diff
@@ -0,0 +1,49 @@
+package semver
```
Comment (erezrokah) [inline cli/internal/semver/semver.go:1]: We use https://github.com/Masterminds/semver in our [backend repo FYI](https://github.com/cloudquery/cloud/blob/08e805674577b347a07b91bc155802c08c1f5bf9/go.mod#L17)

File: cli/cmd/install.go [hunk 5]
```diff
@@ -61,6 +61,10 @@ func installPlugin(cmd *cobra.Command, args []string) error {
 	if err != nil {
 		return fmt.Errorf("failed to get team name: %w", err)
 	}
+
+	pluginVersionWarner, _ := managedplugin.NewPluginVersionWarner(log.Logger, authToken.Value)
+	specs.WarnOnOutdatedVersions(ctx, pluginVersionWarner, sources, destinations, transformers)
```
Comment (disq) [inline cli/cmd/install.go:66]: This will panic on error. Either add a `nil` check here, or a nil check in the method.

Comment (marianogappa) [reply cli/cmd/install.go:66]: Are you sure?
https://github.com/cloudquery/plugin-pb-go/blob/main/managedplugin/version_checker.go#L58

Comment (marianogappa) [reply cli/cmd/install.go:66]: Added the check. But it cannot fail at the moment (hubClient always succeeds). It's still good practice to keep the check.

**Comment by marianogappa** [comment]
That panic in the CI is happening in main btw, still good to look into it but seems unrelated to this PR.

**Review by erezrokah — APPROVED** [review]

**Comment by erezrokah** [comment]
> That panic in the CI is happening in main btw, still good to look into it but seems unrelated to this PR.

The race panic should have been fixed by https://github.com/cloudquery/plugin-pb-go/pull/410 and https://github.com/cloudquery/plugin-pb-go/pull/418

**Comment by marianogappa** [comment]
@erezrokah It's still there 🤔 
Screenshot 2024-10-16 at 17 28 01

**Comment by marianogappa** [comment]
@erezrokah I think you reverted it? It's back on 3.14.6
Screenshot 2024-10-16 at 17 34 35

**Pull request merged**